File: jasaku-backend/src/utils/operating-hours.ts
const OP_START_HOUR = 8;
const OP_START_MIN = 0;
const OP_END_HOUR = 17;
const OP_END_MIN = 0;
const ORDER_CUTOFF_HOUR = 15;
const ORDER_CUTOFF_MIN = 0;
const WARNING_START_HOUR = 14;
const WARNING_START_MIN = 30;

function getTotalMinutes(h: number, m: number): number {
  return h * 60 + m;
}

export function getCurrentWitaTime(): { hour: number; minute: number } {
  const now = new Date();
  const wita = new Date(now.getTime() + 8 * 60 * 60 * 1000);
  return {
    hour: wita.getUTCHours(),
    minute: wita.getUTCMinutes(),
  };
}

export function getTodayWitaDate(): Date {
  const now = new Date();
  const wita = new Date(now.getTime() + 8 * 60 * 60 * 1000);
  return new Date(wita.getUTCFullYear(), wita.getUTCMonth(), wita.getUTCDate());
}

export function isWithinOperatingHours(): boolean {
  const { hour, minute } = getCurrentWitaTime();
  const nowTotal = getTotalMinutes(hour, minute);
  const startTotal = getTotalMinutes(OP_START_HOUR, OP_START_MIN);
  const endTotal = getTotalMinutes(OP_END_HOUR, OP_END_MIN);
  return nowTotal >= startTotal && nowTotal < endTotal;
}

export function canTransitionWorkflow(): { allowed: boolean; message?: string } {
  const { hour, minute } = getCurrentWitaTime();
  const nowTotal = getTotalMinutes(hour, minute);
  const startTotal = getTotalMinutes(OP_START_HOUR, OP_START_MIN);
  const endTotal = getTotalMinutes(OP_END_HOUR, OP_END_MIN);

  if (nowTotal < startTotal) {
    return { allowed: false, message: `Belum jam operasional (${String(OP_START_HOUR).padStart(2, '0')}:${String(OP_START_MIN).padStart(2, '0')} WITA)` };
  }
  if (nowTotal >= endTotal) {
    return { allowed: false, message: `Sudah lewat jam operasional. Batas konfirmasi pekerjaan pukul ${String(OP_END_HOUR - 1).padStart(2, '0')}:59 WITA` };
  }
  return { allowed: true };
}

export function canCompleteWork(): { allowed: boolean; message?: string } {
  const { hour, minute } = getCurrentWitaTime();
  const nowTotal = getTotalMinutes(hour, minute);
  const startTotal = getTotalMinutes(OP_START_HOUR, OP_START_MIN);
  const endTotal = getTotalMinutes(OP_END_HOUR, OP_END_MIN);

  if (nowTotal < startTotal) {
    return { allowed: false, message: `Belum jam operasional (${String(OP_START_HOUR).padStart(2, '0')}:${String(OP_START_MIN).padStart(2, '0')} WITA)` };
  }
  if (nowTotal >= endTotal) {
    return { allowed: false, message: "Sudah lewat jam operasional. Batas konfirmasi selesai pukul 16:59 WITA" };
  }
  return { allowed: true };
}

export function canOrderNow(): { allowed: boolean; warning?: string } {
  const { hour, minute } = getCurrentWitaTime();
  const nowTotal = getTotalMinutes(hour, minute);
  const startTotal = getTotalMinutes(OP_START_HOUR, OP_START_MIN);
  const cutoffTotal = getTotalMinutes(ORDER_CUTOFF_HOUR, ORDER_CUTOFF_MIN);
  const warningStartTotal = getTotalMinutes(WARNING_START_HOUR, WARNING_START_MIN);

  if (nowTotal < startTotal) {
    return { allowed: false, warning: `Belum jam operasional (${String(OP_START_HOUR).padStart(2, '0')}:${String(OP_START_MIN).padStart(2, '0')} WITA)` };
  }
  if (nowTotal >= cutoffTotal) {
    return { allowed: false, warning: "Sudah lewat jam operasional, silahkan order untuk besok" };
  }
  if (nowTotal >= warningStartTotal) {
    return { allowed: true, warning: "Waktu pemesanan mepet dengan jam operasional berakhir. Sarankan order besok pagi jam 08:00 atau lihat jadwal mitra." };
  }
  return { allowed: true };
}

export function isSameWitaDate(date: Date): boolean {
  const now = new Date();
  const nowWita = new Date(now.getTime() + 8 * 60 * 60 * 1000);
  const dateWita = new Date(date.getTime() + 8 * 60 * 60 * 1000);
  return nowWita.getUTCFullYear() === dateWita.getUTCFullYear() &&
         nowWita.getUTCMonth() === dateWita.getUTCMonth() &&
         nowWita.getUTCDate() === dateWita.getUTCDate();
}

File: jasaku_app/lib/core/utils/operating_hours.dart
class OperatingHours {
  static const int startHour = 8;
  static const int startMinute = 0;
  static const int endHour = 16;
  static const int endMinute = 0;
  static const int orderCutoffHour = 15;
  static const int orderCutoffMinute = 59;
  static const int warningStartHour = 15;
  static const int warningStartMinute = 30;

  static int _totalMinutes(int h, int m) => h * 60 + m;

  static DateTime _nowWita() {
    final utc = DateTime.now().toUtc();
    return utc.add(const Duration(hours: 8));
  }

  static bool isWithinOperatingHours() {
    final now = _nowWita();
    final total = _totalMinutes(now.hour, now.minute);
    return total >= _totalMinutes(startHour, startMinute) &&
           total < _totalMinutes(endHour, endMinute);
  }

  static ({bool allowed, String? warning}) canOrderNow() {
    final now = _nowWita();
    final total = _totalMinutes(now.hour, now.minute);
    final start = _totalMinutes(startHour, startMinute);
    final cutoff = _totalMinutes(orderCutoffHour, orderCutoffMinute);
    final warningStart = _totalMinutes(warningStartHour, warningStartMinute);

    if (total < start) {
      return (allowed: false, warning: 'Belum jam operasional (08:00-16:00 WITA)');
    }
    if (total >= cutoff) {
      return (allowed: false, warning: 'Sudah lewat jam operasional, silahkan order untuk besok');
    }
    if (total >= warningStart) {
      return (
        allowed: true,
        warning: 'Jam operasional berakhir pukul 16:00 WITA. Pilih hari lain atau pesan 2 hari kerja. Jika tetap pesan, provider bisa minta extensi jika pekerjaan belum selesai.',
      );
    }
    return (allowed: true, warning: null);
  }

  static bool isToday(DateTime date) {
    final now = _nowWita();
    return date.year == now.year && date.month == now.month && date.day == now.day;
  }
}